crates/red_knot/src/module.rs

```diff
@@ -31,6 +32,55 @@ impl Module {
 
         modules.modules.get(self).unwrap().path.clone()
     }
+
+    pub fn kind<Db>(&self, db: &Db) -> ModuleKind
+    where
+        Db: HasJar<SemanticJar>,
+    {
+        let modules = &db.jar().module_resolver;
+
+        modules.modules.get(self).unwrap().kind
+    }
+
+    pub fn relative_name<Db>(&self, db: &Db, level: u32, module: Option<&str>) -> Option<ModuleName>
+    where
+        Db: HasJar<SemanticJar>,
+    {
+        let name = self.name(db);
+        let kind = self.kind(db);
+
+        let mut components = name.components().peekable();
+
+        if level > 0 {
+            let start = match kind {
+                // `.` resolves to the enclosing package
+                ModuleKind::Module => 0,
+                // `.` resolves to the current package
+                ModuleKind::Package => 1,
+            };
+
+            // Skip over the relative parts.
+            for _ in start..level {
+                components.next_back()?;
+            }
+        }
```

Comment:
I think this is still off-by-one, because ranges are half-open?

If we have a module (of kind module) `foo.bar`, and we an import with level 1, that should resolve to `foo` (strip one level). But this code would resolve it to `foo.bar` (strip nothing), because `0..1` is empty.

This kind of code is where it's really useful to add a few isolated unit tests of the logic, so it's easy to eyeball that it's doing what we expect.

Reply:
Haha and I see you did exactly that! I should read the whole change before commenting, sorry. Let me look at the tests and see where my reading of the code is wrong, or the tests are.

Reply:
Oh of course I'm wrong, 0..1 is not empty, the range is half-open, that has 0 in it. Sorry, clearly haven't woken up yet this morning.

Reviewing the tests found a different issue (level = 0 is not a relative import), but this logic here looks just fine.